python/paddle/distributed/auto_parallel/operators/dist_embedding.py

```diff
@@ -102,118 +102,107 @@ def update_dims_mapping(self, op_dist_attr):
-            copy_distributed_attr_for_dist_op(c_embedding_op, dst_block,
-                                              op_dist_attr)
-            copy_distributed_attr_for_dist_op(c_allreduce_sum_op, dst_block,
-                                              op_dist_attr)
-
-        if in_dygraph_mode():
-            raise NotImplementedError(
-                "Dist op for [{}] with idx [{}] is NOT implemented yet.".format(
-                    "matmul", 0))
+    @staticmethod
+    def forward(ctx, *args, **kwargs):
+        """
+        kwargs: inputname_mapping & outputname_mapping
+        """
+        dst_block = ctx.get_dst_main_program().global_block()
+        src_op = ctx.get_cur_src_op()
+        rank_id = ctx.get_rank_id()
+        op_dist_attr = ctx.get_cur_dist_attr()
+
+        # check validation of inputs / outputs 
+        for input_name in src_op.desc.input_names():
+            assert input_name in kwargs, "input [{}] is not given".format(
+                input_name)
+            assert len(kwargs[input_name]) == len(
+                src_op.desc.input(input_name)
+            ), "number of tensor for input [{}] is not match".format(input_name)
+        for output_name in src_op.desc.output_names():
+            assert output_name in kwargs, "input [{}] is not given".format(
+                output_name)
+            assert len(kwargs[output_name]) == len(
+                src_op.desc.output(output_name)
+            ), "number of tensor for input [{}] is not match".format(
+                output_name)
+
+        Ids_var = dst_block.var(kwargs['Ids'][0])
+        Weight_var = dst_block.var(kwargs['W'][0])
+        Out_var = dst_block.var(kwargs['Out'][0])
+
+        # got dist attribute info
+        embedding_row_dim_mapping = op_dist_attr.get_input_dims_mapping(
+            Weight_var.name)[0]
+        process_mesh_shape = op_dist_attr.get_process_mesh().topology
+        process_mesh_group = op_dist_attr.get_process_mesh().process_group
+
+        # caculate embedding offset
+        # TODO generalize here, using cartisian product to allow any dimensional mesh shape
+        mesh_shape = len(process_mesh_shape)
+        assert mesh_shape <= 2, "row_parallel_embedding only support 1 or 2 dimensional process mesh, but got {}".format(
+            process_mesh_shape)
+        num_partition = process_mesh_shape[embedding_row_dim_mapping]
+        # TODO generalize here, support any mesh group 
+        if mesh_shape == 1:
+            relative_idx = process_mesh_group.index(rank_id)
         else:
-            return static_handle
+            relative_idx = rank_id % num_partition
+
+        per_part_size = Weight_var.shape[0]
+        relative_idx = relative_idx * per_part_size
+
+        # TODO caculate ring id 
+        model_parallel_axis, process_mesh = op_dist_attr.get_owner_context(
+        )._get_model_parallel_info()
+        group_ranks = _get_comm_group(process_mesh.process_group,
```

Comment:
Can the  _get_comm_group function become a member function of ProcessMesh？Since the first arguments are related to the same ProcessMesh object.